stl/inc/expected

```diff
@@ -0,0 +1,1034 @@
+// expected standard header
+
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#pragma once
+#ifndef _EXPECTED_
+#define _EXPECTED_
+#include <yvals.h>
+#if _STL_COMPILER_PREPROCESSOR
+#ifndef __cpp_lib_expected
+#pragma message("The contents of <expected> are available only with C++23 or later.")
+#else // ^^^ !__cpp_lib_expected / __cpp_lib_expected vvv
+#include <concepts>
+#include <exception>
+#include <initializer_list>
+#include <memory>
+#include <type_traits>
+#include <xutility>
+
+#pragma pack(push, _CRT_PACKING)
+#pragma warning(push, _STL_WARNING_LEVEL)
+#pragma warning(disable : _STL_DISABLED_WARNINGS)
+_STL_DISABLE_CLANG_WARNINGS
+#pragma push_macro("new")
+#undef new
+
+_STD_BEGIN
+
+// [expected.un.object]
+template <class _Err>
+class unexpected {
+    static_assert(is_object_v<_Err>, "E must be an object type");
+    static_assert(!is_array_v<_Err>, "E must not be an array type");
+    static_assert(!is_volatile_v<_Err> && !is_const_v<_Err>, "E must not be cv-qualified");
+    static_assert(!_Is_specialization_v<_Err, unexpected>, "E must not be a specialization of unexpected");
+
+    template <class _Ty, class _Err2>
+    friend class expected;
+
+public:
+    // [expected.un.ctor]
+    template <class... _Args>
+        requires is_constructible_v<_Err, _Args...>
+    constexpr explicit unexpected(in_place_t, _Args&&... _Vals) noexcept(
+        is_nothrow_constructible_v<_Err, _Args...>) // strengthened
+        : _Unexpected(_STD forward<_Args>(_Vals)...) {}
+
+    // clang-format off
+    template <class _Uty, class... _Args>
+        requires is_constructible_v<_Err, initializer_list<_Uty>&, _Args...>
+    constexpr explicit unexpected(in_place_t, initializer_list<_Uty> _Ilist, _Args&&... _Vals) noexcept(
+               is_nothrow_constructible_v<_Err, initializer_list<_Uty>&, _Args...>) // strengthened
+        : _Unexpected(_Ilist, _STD forward<_Args>(_Vals)...) {}
+
+    template <class _UError = _Err>
+        requires (!is_same_v<remove_cvref_t<_UError>, unexpected> && !is_same_v<remove_cvref_t<_UError>, in_place_t>
+                 && is_constructible_v<_Err, _UError>)
+    constexpr explicit unexpected(_UError&& _Unex) noexcept(is_nothrow_constructible_v<_Err, _UError>) // strengthened
+        : _Unexpected(_STD forward<_UError>(_Unex)) {}
+    // clang-format on
+
+    // [expected.un.obs]
+#ifdef __cpp_explicit_this_parameter // TRANSITION, P2549, P0847R7
+    _NODISCARD constexpr decltype(auto) error(this auto&& _Self) noexcept {
+        return (_STD forward<decltype(_Self)>(_Self)._Unexpected);

```

Comment:
@CaseyCarter notes that this may not compile in the scenario `class Derived : private expected<T, E> { };` I think (haven't confirmed) that access control will prevent this. Casey thinks this is a Core Language defect, but since this is not necessary to achieve `<expected>`'s goals, I recommend simply dropping all attempts to use `__cpp_explicit_this_parameter` for now.

Reply:
To work correctly on types that derive from `expected` (possibly with protected or private inheritance) I believe this needs to be:
```c++
_NODISCARD constexpr auto&& error(this auto&& _Self) noexcept {
    return ((_Like_t<decltype(_Self), expected>>)_Self)._Unexpected; // C-style cast needed to bypass access control
}
```
where `_Like_t` is:
```c++
template <class _Source, class _Target>
using _Like_t = typename _Add_qualifiers<_Source>::template _Apply<_Target>;
```
reusing `_Add_qualifiers` from the implementation of `common_reference`. 

The C-style cast makes this super icky. How do folks feel about wrapping the mess in a function template a la:
```c++
template <class _Raw, class _Source>
_NODISCARD constexpr auto&& _Forward_as(auto&& _Val) noexcept {
    _STL_INTERNAL_STATIC_ASSERT(same_as<_Source&, decltype(_Val)&>);
    using _Cooked = typename _Add_qualifiers<_Source&&>::template _Apply<_Raw>;
    return (_Cooked) _Val; // C-style cast used to bypass access control
}
```
so callers can be nice and clean:
```c++
template <class _SelfTy>
_NODISCARD constexpr auto&& error(this _SelfTy&& _Self) noexcept {
    return _Forward_as<expected, _SelfTy>(_Self)._Unexpected;
}
```
does this contain the ugliness well enough for us to be able to use explicit object functions in the STL?

Reply:
I am not too happy with the increased throughput cost. But I would expect that if it ever became significant somebody in the frontend team would have their magic wand.